broker/src/main/java/org/apache/rocketmq/broker/transaction/rocksdb/TransactionalMessageRocksDBService.java

```diff
@@ -0,0 +1,271 @@
+import org.apache.rocketmq.store.MessageStore;
+import org.apache.rocketmq.store.rocksdb.MessageRocksDBStorage;
+import org.apache.rocketmq.store.transaction.TransRocksDBRecord;
+import org.apache.rocketmq.store.transaction.TransMessageRocksDBStore;
+import static org.apache.rocketmq.store.rocksdb.MessageRocksDBStorage.TRANS_COLUMN_FAMILY;
+
+public class TransactionalMessageRocksDBService {
+    private static final Logger log = LoggerFactory.getLogger(LoggerName.TRANSACTION_LOGGER_NAME);
+    private static final int MAX_BATCH_SIZE_FROM_ROCKSDB = 2000;
+    private static final int INITIAL = 0, RUNNING = 1, SHUTDOWN = 2;
+    private volatile int state = INITIAL;
+
+    private final MessageRocksDBStorage messageRocksDBStorage;
+    private final TransMessageRocksDBStore transMessageRocksDBStore;
+    private final MessageStore messageStore;
+    private final BrokerController brokerController;
+
+    private TransStatusCheckService transStatusService;
+    private ExecutorService checkTranStatusTaskExecutor;
+
+    public TransactionalMessageRocksDBService(final MessageStore messageStore, final BrokerController brokerController) {
+        this.messageStore = messageStore;
+        this.transMessageRocksDBStore = messageStore.getTransRocksDBStore();
+        this.messageRocksDBStorage = transMessageRocksDBStore.getMessageRocksDBStorage();
+        this.brokerController = brokerController;
+    }
+
+    public void start() {
+        if (this.state == RUNNING) {
+            return;
+        }
+        initService();
+        this.transStatusService.start();
+        this.state = RUNNING;
+        log.info("TransactionalMessageRocksDBService start success");
+    }
+
+    private void initService() {
+        this.transStatusService = new TransStatusCheckService();
+        this.checkTranStatusTaskExecutor = ThreadUtils.newThreadPoolExecutor(
+            2,
+            5,
+            100,
+            TimeUnit.SECONDS,
+            new ArrayBlockingQueue<>(2000),
+            new ThreadFactoryImpl("Transaction-rocksdb-msg-check-thread", brokerController.getBrokerIdentity()),
+            new CallerRunsPolicy());
+    }
+
+    public void shutdown() {
+        if (this.state != RUNNING || this.state == SHUTDOWN) {
+            return;
+        }
+        if (null != this.transStatusService) {
+            this.transStatusService.shutdown();
+        }
+        if (null != this.checkTranStatusTaskExecutor) {
+            this.checkTranStatusTaskExecutor.shutdown();
+        }
+        this.state = SHUTDOWN;
+        log.info("TransactionalMessageRocksDBService shutdown success");
+    }
+
+    private void checkTransStatus() {
+        long count = 0;
+        byte[] lastKey = null;
+        while (true) {
+            try {
+                List<TransRocksDBRecord> trs = messageRocksDBStorage.scanRecordsForTrans(TRANS_COLUMN_FAMILY, MAX_BATCH_SIZE_FROM_ROCKSDB, lastKey);
+                if (null == trs || CollectionUtils.isEmpty(trs)) {
```

Comment:
The `null == trs` check is redundant.
```suggestion
                if (CollectionUtils.isEmpty(trs)) {
```

Reply:
you are right ~